File: C++/103. Binary Tree Zigzag Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> sol;
        if(root==NULL){
            return sol;
        }
        queue<TreeNode*> q;
        q.push(root);
        bool flag=true;
        while(!q.empty()){
            int n=q.size();
            vector<int> vec(n);
            for(int i=0;i<n;i++){
                TreeNode* curr=q.front();
                q.pop();
                vec[i]=curr->val;
                if(curr->left){
                    q.push(curr->left);
                }
                if(curr->right){
                    q.push(curr->right);
                }
            }
            if(!flag){
                reverse(vec.begin(),vec.end());
            }
            flag=!flag;
            sol.push_back(vec);
        }
        return sol;
    }
};

File: Java/103. Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> sol=new LinkedList<>();
        Queue<TreeNode> q=new LinkedList<>();
        if(root==null){
            return sol;
        }
        q.offer(root);
        int count = 1;
        while(!q.isEmpty()){
            int n=q.size();
            List<Integer> subL=new LinkedList<>();
            for(int i=0;i<n;i++){
                if(q.peek().left!=null){
                    q.offer(q.peek().left);
                }
                if(q.peek().right!=null){
                    q.offer(q.peek().right);
                }
                subL.add(q.poll().val);
            }
            if(count%2==0){
                Collections.reverse(subL);
            }
            count++;
            sol.add(subL);
        }
        return sol;
    }
}
